R/fimsfit.R

```diff
@@ -0,0 +1,204 @@
+                     control=NULL, save.sdrep=FALSE,
+                     filename=NULL){
+if(!is.null(input$random)) {
+  cli::cli_abort("Random effects declared but not implemented yet.")
+}
+if(newtonsteps>0) {
+  cli::cli_abort("Newton steps not implemented yet.")
+}
+if (loopnum < 0) {
+  cli::cli_abort("loopnum ({.par {loopnum}}) must be >= 0.")
+}
+obj <- MakeADFun(data=list(), parameters=input$parameters,
+                 map=input$map, random=input$random,
+                 DLL='FIMS', silent=TRUE)
+if(!do.fit) return(list(obj=obj, rep=obj$report(obj$env$last.par.best)))
+# to do: max this update elements that are not supplied by default
+if(is.null(control))
+  control <- list(eval.max=10000, iter.max=10000, trace=0)
+if(!is_fims_verbose()) control$trace <- 0
+## optimize and compare
+t0 <- Sys.time()
+cli::cli_inform(c("v" = "Starting optimization ..."))
+opt0 <- opt <-
+  with(obj, nlminb(start = par, objective = fn, gradient = gr, control=control))
+maxgrad0 <- maxgrad <- max(abs(obj$gr(opt$par)))
+if(loopnum>0){
+  cli::cli_inform(c(
+    "i" = "Restarting optimizer {loopnum} times silently to improve gradient."
+  ))
+    for(ii in 2:loopnum){
+      # TODO: Shouldn't this be dictated by verbose?
+      control$trace <- 0
+    opt <- with(obj, nlminb(start = opt$par, objective = fn,
+                            gradient = gr, control=control))
+    maxgrad <- max(abs(obj$gr(opt$par)))
+  }
+  div_digit <- cli::cli_div(theme = list(.val = list(digits = 5)))
+  cli::cli_inform(c(
+    "i" = "Maximum gradient went from {.val {maxgrad0}} to
+          {.val {maxgrad}} after {loopnum} steps."
+  ))
+  cli::cli_end(div_digit)
+}
+n_total <- length(obj$env$last.par.best)
+n_fe <- length(obj$par)
+opt$num_pars <- list(total=n_total, fixed_effects=n_fe, random_effects=n_total-n_fe)
+# TODO: This if statement should be in the constructor of the class
+if(is.null(input$version)) {
+  default_version <- "FIMS model"
+  cli::cli_warn(c(
+    "i" = "No {.par input$version} string provided,
+      using default of {.val {default_version}}."
+  ))
+  input$version <- default_version
+}
+time_optimization <- Sys.time() - t0
+cli::cli_inform(c("v" = "Finished optimization"))
+opt$max_gradient <- maxgrad
+
+rep <- c(version=input$version, obj$report())
+sdrep <- std <- NULL
+time_sdreport <- NA
+if(getsd){
+  t2 <- Sys.time()
+  sdrep <- TMB::sdreport(obj)
+  std <- summary(sdrep)
+  std <- data.frame(dimnames(std)[[1]], std)
+  names(std) <- c('name', 'est', 'se')
+  std$lwr <- std$est - 1.96*std$se
+  std$upr <- std$est + 1.96*std$se
+  row.names(std) <- NULL
+  cli::cli_inform(c("v" = "Finished sdreport"))
+  time_sdreport <- Sys.time() - t2
+}
+parList <- obj$env$parList()
+parnames <- names(obj$par)
+parnames <- as.vector((unlist(sapply(unique(parnames), function(x){
+  temp <- parnames[parnames==x]
+  if(length(temp)>1) paste0(temp,'[',1:length(temp),']') else temp
+}))))
```

Comment:
I don't mind if you change it, but note that your case cannot happen b/c TMB would never do that. It flattens a named list so you'd never have other names inserted in a vector like that. The original works for me on models with >1 parameter. 

Eventually this would be replaced with a function that returns the actual parameter names I think.